### PR TITLE
Bintray publish settings

### PR DESCRIPTION
**Pull request by jamescway**
### Problem

We need to release artifacts built in travis to Bintray openzipkin/zipkin repo.
#### How it works
- user creates a tag `git tag v1.2.3` 
- then does `git push origin master --tags`
- triggers the publish script to be called after travis build is successful
- script creates a bintray credentials file in travis... 
- then using bintray-sbt pushes artifacts to packages/modules in bintray (i.e. openzipkin/zipkin/zipkin-collector-core)
  - based on encrypted values in travis.yml
  - `gem install travis`
  - `travis encrypt BINTRAY_API_KEY=<Bintray API key>--add`
  - `travis encrypt BINTRAY_USER="<Bintray username" --add`
### Caveats
- not detecting if the tagged branch is master due to bug in travis (https://github.com/travis-ci/travis-ci/issues/1675)
- still running tests when tag is pushed, needs workaround
- modules/packages in Openzipkin repo need to be created manually (i.e. zipkin-collector-core) due to https://github.com/softprops/bintray-sbt/issues/58
- current version number in Project.scala
- building from oraclejdk7

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,35 +1,31 @@
 language: scala
```
Comment (codefromthecrypt) [inline .travis.yml:1]: supernit: can you get rid of the fuzz here? looks like the only needed one is the last few lines.

File: project/Project.scala [hunk 2]
```diff
@@ -85,6 +87,15 @@ object Zipkin extends Build {
       "org.apache.zookeeper" % "zookeeper" % "3.4.6", // internal twitter + kafka + curator
       "org.slf4j" % "slf4j-api" % "1.6.4" // libthrift 0.5 otherwise pins 1.5.x
     )
+    // Stop publishing to local repo
```
Comment (codefromthecrypt) [inline project/Project.scala]: `publish-local` does this right?

File: project/plugins.sbt [hunk 3]
```diff
@@ -9,3 +9,12 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-site" % "0.8.1")
 addSbtPlugin("com.twitter" %% "scrooge-sbt-plugin" % "3.16.3")
 
 addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.9.2")
+
+resolvers += Resolver.url(
```
Comment (codefromthecrypt) [inline project/plugins.sbt]: rebase against master, as we have this, now

File: project/plugins.sbt [hunk 4]
```diff
@@ -9,3 +9,12 @@ addSbtPlugin("com.typesafe.sbt" % "sbt-site" % "0.8.1")
...
+"bintray-sbt-plugin-releases",
+url("http://dl.bintray.com/content/sbt/sbt-plugin-releases"))(
+Resolver.ivyStylePatterns)
+
+addSbtPlugin("me.lessis" % "bintray-sbt" % "0.3.0")
+
+resolvers += Classpaths.sbtPluginReleases
```
Comment (codefromthecrypt) [inline project/plugins.sbt]: remove this as it is redundant.. I think all you need is bintray-sbt..

File: travis/publish.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+echo "********************************"
+echo "***** Publishing Artifacts *****"
+echo "********************************"
+echo "Branch: ${TRAVIS_BRANCH}"
+echo "Tag: ${TRAVIS_TAG}"
+echo "JDK Version: ${TRAVIS_JDK_VERSION}"
+
+#Weird comparison comparing branch to tag because when you 'git push --tags'
```
Comment (codefromthecrypt) [inline travis/publish.sh]: great comment!

File: project/Project.scala [hunk 6]
```diff
@@ -17,9 +17,11 @@
 import io.zipkin.sbt.{BuildProperties,PackageDist,GitProject}
 import sbt._
 import Keys._
+import bintray.BintrayKeys._
 
 object Zipkin extends Build {
-  val zipkinVersion = "1.2.0-SNAPSHOT"
+
+  val zipkinVersion = "1.2.0-ALPHA"
```
Comment (codefromthecrypt) [inline project/Project.scala]: This should stay snapshot. The release number should be taken from the tag name.

File: travis/publish.sh [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+#the branch somehow becomes the tag value
+#github issue: https://github.com/travis-ci/travis-ci/issues/1675
+test "${TRAVIS_BRANCH}" = "${TRAVIS_TAG}" &&
```
Comment (codefromthecrypt) [inline travis/publish.sh]: maybe use something more like the if statement here, as it emits what's going on? https://github.com/Netflix/feign/blob/master/installViaTravis.sh

Comment (codefromthecrypt) [inline travis/publish.sh]: or here https://github.com/Netflix/feign/blob/master/buildViaTravis.sh

File: travis/publish.sh [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+test "${TRAVIS_TAG}" != "" &&
+test "${TRAVIS_JDK_VERSION}" = 'oraclejdk7' &&
+sh "$TRAVIS_BUILD_DIR/.travis-ci-bintray-credentials.sh" &&
+sbt 'set version := version.value + "." + System.getenv("TRAVIS_BUILD_NUMBER")' publish
```
Comment (codefromthecrypt) [inline travis/publish.sh]: I think this is where you want to parse the tag to get the version number, or use a plugin that does that for you.

**Comment by codefromthecrypt** [comment]
Thanks for giving this a go, @jamescway! I think once you rebase against upstream master, this gets smaller. I've made some notes, mostly on cleanup.

File: project/Project.scala (hunk 2 above)
Comment (jamescway) [inline project/Project.scala]: yeah, now that you mention that, this seems totally unnecessary and seems to override expected behavior, removing altogether.

**Comment by jamescway** [comment]
having some problems after rebase...
traced it down to this line 
`addSbtPlugin("com.twitter" %% "scrooge-sbt-plugin" % "3.19.0")`
previously it was
`addSbtPlugin("com.twitter" %% "scrooge-sbt-plugin" % "3.16.3")`

With the newer lib `sbt publish` now causes: 
`[error] (zipkin-collector-core/*:publish) java.lang.NoSuchMethodError: org.jboss.netty.handler.codec.http.HttpRequest.setHeader(Ljava/lang/String;Ljava/lang/Object;)V
`

here's the full stack trace

```
> last zipkin-collector-core/*:publish
java.lang.NoSuchMethodError: org.jboss.netty.handler.codec.http.HttpRequest.setHeader(Ljava/lang/String;Ljava/lang/Object;)V
    at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.construct(NettyAsyncHttpProvider.java:693)
    at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.buildRequest(NettyAsyncHttpProvider.java:650)
    at com.ning.http.client.providers.netty.NettyConnectListener$Builder.build(NettyConnectListener.java:144)
    at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.doConnect(NettyAsyncHttpProvider.java:1070)
    at com.ning.http.client.providers.netty.NettyAsyncHttpProvider.execute(NettyAsyncHttpProvider.java:935)
    at com.ning.http.client.AsyncHttpClient.executeRequest(AsyncHttpClient.java:499)
    at dispatch.HttpExecutor$class.apply(execution.scala:47)
    at dispatch.Http.apply(execution.scala:12)
    at dispatch.HttpExecutor$class.apply(execution.scala:42)
    at dispatch.Http.apply(execution.scala:12)
    at bintry.Requests.request(Client.scala:42)
    at bintry.Methods$Repo$Package$Version$Upload.apply(Methods.scala:177)
    at bintray.BintrayIvyRepository.put(Resolver.scala:54)
    at org.apache.ivy.plugins.resolver.RepositoryResolver.put(RepositoryResolver.java:234)
    at org.apache.ivy.plugins.resolver.RepositoryResolver.publish(RepositoryResolver.java:216)
    at sbt.IvyActions$$anonfun$publish$3.apply(IvyActions.scala:336)
    at sbt.IvyActions$$anonfun$publish$3.apply(IvyActions.scala:335)
    at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:772)
    at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
    at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
    at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:771)
    at sbt.IvyActions$.publish(IvyActions.scala:335)
    at sbt.IvyActions$$anonfun$publish$1$$anonfun$apply$1.apply$mcV$sp(IvyActions.scala:121)
    at sbt.IvyActions$$anonfun$publish$1$$anonfun$apply$1.apply(IvyActions.scala:121)
    at sbt.IvyActions$$anonfun$publish$1$$anonfun$apply$1.apply(IvyActions.scala:121)
    at sbt.IvyActions$.withChecksums(IvyActions.scala:130)
    at sbt.IvyActions$.sbt$IvyActions$$withChecksums(IvyActions.scala:125)
    at sbt.IvyActions$$anonfun$publish$1.apply(IvyActions.scala:121)
    at sbt.IvyActions$$anonfun$publish$1.apply(IvyActions.scala:114)
    at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:155)
    at sbt.IvySbt$Module$$anonfun$withModule$1.apply(Ivy.scala:155)
    at sbt.IvySbt$$anonfun$withIvy$1.apply(Ivy.scala:132)
    at sbt.IvySbt.sbt$IvySbt$$action$1(Ivy.scala:57)
    at sbt.IvySbt$$anon$4.call(Ivy.scala:65)
    at xsbt.boot.Locks$GlobalLock.withChannel$1(Locks.scala:93)
    at xsbt.boot.Locks$GlobalLock.xsbt$boot$Locks$GlobalLock$$withChannelRetries$1(Locks.scala:78)
    at xsbt.boot.Locks$GlobalLock$$anonfun$withFileLock$1.apply(Locks.scala:97)
    at xsbt.boot.Using$.withResource(Using.scala:10)
    at xsbt.boot.Using$.apply(Using.scala:9)
    at xsbt.boot.Locks$GlobalLock.ignoringDeadlockAvoided(Locks.scala:58)
    at xsbt.boot.Locks$GlobalLock.withLock(Locks.scala:48)
    at xsbt.boot.Locks$.apply0(Locks.scala:31)
    at xsbt.boot.Locks$.apply(Locks.scala:28)
    at sbt.IvySbt.withDefaultLogger(Ivy.scala:65)
    at sbt.IvySbt.withIvy(Ivy.scala:127)
    at sbt.IvySbt.withIvy(Ivy.scala:124)
    at sbt.IvySbt$Module.withModule(Ivy.scala:155)
    at sbt.IvyActions$.publish(IvyActions.scala:114)
    at sbt.Classpaths$$anonfun$publishTask$1.apply(Defaults.scala:1283)
    at sbt.Classpaths$$anonfun$publishTask$1.apply(Defaults.scala:1282)
    at scala.Function3$$anonfun$tupled$1.apply(Function3.scala:35)
    at scala.Function3$$anonfun$tupled$1.apply(Function3.scala:34)
    at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
    at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
    at sbt.std.Transform$$anon$4.work(System.scala:63)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
    at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
    at sbt.Execute.work(Execute.scala:235)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
    at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
    at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
```

Seems like maybe the lib [bintry](https://github.com/softprops/bintry) is using an old version of [dispatch](http://dispatch.databinder.net/Dispatch.html) maybe?

**Comment by jamescway** [comment]
looks like we have netty-3.10.1.Final.jar and async-http-client-1.8.10.jar both loading netty

**Comment by eirslett** [comment]
Where does async-http-client come from? Should be easy to replace with finagle-http if we have control over the source.

**Comment by eirslett** [comment]
Aah, I think I can spot the problem - publish to bintray uses Dispatch, while Scrooge uses Finagle (with an older version of Netty, probably)
I thought sbt used separate class loaders for each plugin, maybe not...

**Comment by codefromthecrypt** [comment]
Could an exclude on scrooge work? I doubt the compiler actually needs to
load the netty classes. I can give this a try in 30m or so, if no one else
does.

**Comment by jamescway** [comment]
@eirslett yeah, dispatch uses it, I have a [PR](https://github.com/softprops/bintry/pull/15) merged into bintry going to find out if he will [release](https://github.com/softprops/bintry/issues/16) it.  Actually scrooge has the newer version of netty, bintry (dep of bintray-sbt has an older version.

@adriancole rolling back and using scrooge-sbt-plugin" % "3.16.3" would solve the problem in the meantime

What do u guys think?  Going to roll it back to 3.16.3 feel free to thumbs down :p

**Comment by codefromthecrypt** [comment]
> @adriancole rolling back and using scrooge-sbt-plugin" % "3.16.3" would
> solve the problem in the meantime
> I'd be fine with adding a commit to use this version with a comment
> about why including the link to the fix. However, it looks like the
> commit didn't change any source files. Another way could be to make a
> commit that overrides the version of dispatch-json4s-native. This
> helps as we already need to update scrooge, and mixed classpath seems
> worse. wdyt?

**Comment by rtyler** [comment]
This probably isn't going to work now that #503 has been merged <_<

**Comment by abesto** [comment]
OTOH Bintray publishes an official Gradle plugin: https://bintray.com/jfrog/jfrog-jars/gradle-bintray-plugin

**Comment by jamescway** [comment]
After update to use gradle https://github.com/openzipkin/zipkin/pull/503, much of this stuff is not relevant, going to close and start another PR

**Comment by jamescway** [comment]
@abesto just noticed that tool, I'll give it a shot

**Pull request closed without merging**